### PR TITLE
Fix DataSelector not working in full app embedding

### LINKED ISSUE #21511: Data selector not working in FullApp embedding if going directly from location that hasn't loaded `/api/database` yet
**Describe the bug**
If FullApp embedding directly to e.g `/collection/root`, then it is not possible to click the :heavy_plus_sign: (New...) menu to start a question, because there hasn't been any calls to `/api/database` yet.

**Possible workaround:**
Create a button outside the iframe, similar to pre-42, which changes the URL to `/question/new` or `/question/notebook#eyJjcmVhdGlvblR5cGUiOiJjdXN0b21fcXVlc3Rpb24iLCJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjpudWxsLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOm51bGx9LCJ0eXBlIjoicXVlcnkifSwiZGlzcGxheSI6InRhYmxlIiwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319`
Or use the homepage as starting point, which loads the databases.

**Reproduce 1**
1. Setup an instance. Might need to adjust allowed embed URLs or just `*`
2. Create a page that has an iframe like (some path that doesn't load the `/api/database` endpoint, which the homepage does by default)
`<iframe src="https://metabase.example.com/collection/root">`
3. Click :heavy_plus_sign: (New...) > Question
image
4. Data selector just shows: `To pick some data, you'll need to add some first`
image

**Reproduce 2**
1. Setup an instance. Might need to adjust allowed embed URLs or just `*`
2. Create a page that has an iframe like (some path that doesn't load the `/api/database` endpoint, which the homepage does by default)
`<iframe src="https://metabase.example.com/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgMSIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjF9LCJkaXNwbGF5Ijoic2NhbGFyIiwiZGlzcGxheUlzTG9ja2VkIjp0cnVlLCJwYXJhbWV0ZXJzIjpbXSwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319">`
3. It will not show the database selector
image
4. Clicking the Reference (Learn about your data) opens the sidebar and makes a request to `/api/database`, which then updates everything correctly
image

**Reproduce X**
There's likely multiple other ways of reproducing this problem, not going through all of them.

**Information about your Metabase Installation:**
Tested 1.42.3

**Additional context**
Historically related to #18552 and #21055, and likely also related to #10927

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/21511
Reproduces https://github.com/metabase/metabase/issues/21511

How to test:
- Reproductions rely on specific flags & user permissions. I highly recommend relying on new cypress tests.
- New tests should fail without code changes and pass otherwise.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/28097)

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/DataSelector/DataSelector.jsx [hunk 1]
```diff
@@ -193,25 +193,32 @@ const DataSelector = _.compose(
     loadingAndErrorWrapper: false,
   }),
   connect(
-    (state, ownProps) => ({
-      metadata: getMetadata(state),
-      databases:
-        ownProps.databases ||
+    (state, ownProps) => {
+      const databases =
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/DataSelector/DataSelector.jsx:197]: He we extract `databases` to a variable and use in `getHasDataAccess`

File: frontend/src/metabase/selectors/data.ts [hunk 2]
```diff
@@ -1,25 +1,18 @@
-import { createSelector } from "reselect";
 import { getEngineNativeType } from "metabase/lib/engine";
-import { State } from "metabase-types/store";
+import { Database } from "metabase-types/api";
 
-const getDatabaseList = createSelector(
-  (state: State) => state.entities.databases,
-  databases => (databases ? Object.values(databases) : []),
-);
+export const getHasDataAccess = (databases: Database[]) => {
```
Comment (ranquild) [inline frontend/src/metabase/selectors/data.ts:4]: Rely on `databases` passed explicitly to force using entity loaders

File: frontend/src/metabase/selectors/data.unit.spec.ts [hunk 3]
```diff
@@ -1,120 +1,99 @@
...
-      const state = createMockState({
-        entities: {
-          databases: {
-            0: { ...mockedDatabase, is_saved_questions: false },
-          },
-        },
-      });
-
-      expect(getHasDataAccess(state)).toBe(true);
+      const databases = [
```
Comment (ranquild) [inline frontend/src/metabase/selectors/data.unit.spec.ts:12]: Pass `databases` directly

File: frontend/src/metabase/query_builder/components/DataSelector/DataSelector.jsx [hunk 4]
```diff
@@ -182,6 +182,10 @@ class DataSelectorInner extends Component {
 }
 
 const DataSelector = _.compose(
+  Databases.loadList({
+    loadingAndErrorWrapper: false,
+    listName: "allDatabases",
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/DataSelector/DataSelector.jsx:187]: There is existing `databases` prop that should not be overwritten

**Comment by kulyk** [comment]
@ranquild the changes look great! I think the refactored functions in `metabase/selectors/data` can't actually be called selectors though as they're not dealing with the redux state anymore. Should we consider moving them elsewhere? Maybe something like `metabase/databases/utils`? I believe we shouldn't call them in `mapStateToProps` anymore as well

**Comment by ranquild** [comment]
@kulyk Agree I can move them to utils. As for not calling them in mapStateToProps, I would rather not do that to be able to backport this PR

**Review by kulyk — APPROVED** [review]
@ranquild I think it's a good idea to merge with the smallest possible diff, so we can backport easily. It'd be cool if we can make it beautiful in a subsequent PR 👍

**Comment by paoliniluis** [comment]
@ranquild remember to backport this manually to 45

**Pull request merged**